### DISCUSSION TITLE
Unity "auto-install" on Windows + with caching

### DISCUSSION THREAD

**Discussion by MinaPecheux** [codemagic-ci-cd/codemagic-docs#910] [Q&A] [answered]
Hi everyone! :slightly_smiling_face:
I had some questions about the new super cool ["auto-install Unity" feature](https://docs.codemagic.io/knowledge-base/install-unity-version/). I've copied the code from the docs in my `codemagic.yaml` but I can't seem to have it work properly in my workflows, with the caching and all :sweat_smile:

- **on Mac:** I think the unity version is not cached: when the build finished, the cache that is packaged is just ~40 MB, so it clearly doesn't contain the unity version. I believe that's because I try to auto-extract the `UNITY_VERSION` from the `ProjectVersion.txt` file, and so it is not readily available at the start when the pipeline looks at the cache paths, but I'm not sure exactly when those cache paths are checked so perhaps I'm wrong ^^ (👉 [build](https://codemagic.io/app/62e0ec9f097606f227bb662f/build/6306357650a2adf052ecf7ea))

- **on Windows:** the workflow crashes when I try to activate the license before installing my new unity version. Usually, to have the license activation work on Windows, I actually needed to do it at the same time as the build... I'm not sure how to separate it from the build step to put it before the install :wink:  (👉 [build](https://codemagic.io/app/62e0ec9f097606f227bb662f/build/630634fb50a2adc3dd896103))

Do you have any idea/advice?
Thanks in advance! :slightly_smiling_face:

**Comment by mohammedbabelly20** [comment]
Hey @MinaPecheux 🙂

Regarding the **Mac** issue: it seems like the cache path not reading the new `UNITY_VERSION` value! no sure if it will help to create a variable at the vars section with that name with a default value, need to check it first.

About the **Windows** build; I think there were no available seats at the time for the licenses! maybe you can try to build again now.

**Reply by MinaPecheux** [reply]
@codemagic-mohammed Hi, thanks for looking into it! :)
About the **Windows** build: no luck retrying. My [new Win build](https://codemagic.io/app/62e0ec9f097606f227bb662f/build/63063f3afd736abfb166f6b8) crashes at the same step :(

**Comment by mikhail-tokarev** [comment]
I don't think approach with dynamic cache path using `UNITY_VERSION` will work because cache restoring happens before any custom scripts. So you need to specify static cache path which makes `UNITY_VERSION` is redundant here

**Reply by MinaPecheux** [reply]
Thanks for the insight @mikhail-tokarev! :)
So I can't actually "restore" my Unity version from the `ProjectVersion.txt` if i want to cache it, if I understand properly? I have to use a user-defined env var directly in my `codemagic.yaml`?

**Reply by mikhail-tokarev** [reply]
correct, you need to define environment variable **before** build or use static cache path.

**Reply by MinaPecheux** [reply]
Got it! Thank you 👍

**Comment by mikhail-tokarev** [comment, marked as answer]
re Windows error: I can see you use `$UNITY_EMAIL` when activating license but you have only `$UNITY_USERNAME` in environment variables. In this case the error message looks correct
```
UnityConnectLoginRequest: Failed to login - please check your username or password
No sufficient permissions while processing request "https://core.cloud.unity3d.com/api/login", HTTP error code 401
```

**Reply by MinaPecheux** [reply]
Wow, nice catch @mikhail-tokarev! Indeed, it solves the problem and the build is running fine now :)
Just a note then: I copied this code directly from [the new docs](https://docs.codemagic.io/knowledge-base/install-unity-version/), which probably means there is a little inconsistency between the Mac and Windows version... ;)

Thanks to you both for your help!

**Reply by mohammedbabelly20** [reply]
@MinaPecheux We're trying to use the `$UNITY_EMAIL` variable name because it makes more sense.

**Reply by MinaPecheux** [reply]
@codemagic-mohammed Yup, I think you're right! But then perhaps it would be better to update the Mac version in the docs to also use the `UNITY_EMAIL` ;) (just so you can copy-paste the sample without any trouble with env vars ^^)

**Reply by mohammedbabelly20** [reply]
Will keep that in mind, thanks @MinaPecheux.

FYI @todah-zg.
